### ISSUE TITLE
AzureFileCopy@4 to VM with WinRM prerequisites is creating invalid Expired Certificate (1/1/2022)

### ISSUE THREAD

**Issue by marcusrichards** [microsoft/azure-pipelines-tasks#15656]
## Note  
Issues in this repo are for tracking bugs, feature requests and questions for the tasks in this repo 

For a list:  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

If you have an issue or request for the Azure Pipelines service, use developer community instead:  

https://developercommunity.visualstudio.com/spaces/21/index.html )

## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: AzureFileCopy@4

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - Azure Pipelines
    - If using Azure Pipelines, provide the account name: sword-grc, 
    - team project name: ARM
    - build definition name: BranchTestPrototype 
    - build number: 60300
   
[Build Link](https://dev.azure.com/sword-grc/ARM/_build/results?buildId=60300&view=results)

- Agent - Hosted: 
    - If using Hosted agent, provide agent queue name: Azure Pipelines

## Issue Description
Self-Certified Certificate created on VM by FileCopy prerequisites has expiry date of 1/1/2022, hence is invalid and so WinRM copy fails with message:
```
"The remote session query failed for 13.87.91.35 with the following error message: The server certificate on the destination computer (13.87.91.35:5986) has the following errors:      
The SSL certificate is expired."
```

**Explanation:**
[ConfigureWinRM.ps1](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/501dc7d24537e820df7c80bce51aba9674233b2b/201-vm-winrm-windows/ConfigureWinRM.ps1) has code which erroneously hardwires the expiry date of the certificate as follows:

```
$serial = Get-Random
    .\makecert -r -pe -n CN=$hostname -b 01/01/2012 -e **01/01/2022** -eku 1.3.6.1.5.5.7.3.1 -ss my -sr localmachine -sky exchange -sp "Microsoft RSA SChannel Cryptographic Provider" -sy 12 -# $serial 2>&1 | Out-Null
```

![ConfigureWinRM](https://user-images.githubusercontent.com/34490856/148142883-806f8fdb-29bd-483d-a6a4-94b8d896a135.PNG)

### Task logs

[Enable debug logging and please provide the zip file containing all the logs for a speedy resolution]
2022-01-04T17:51:34.2640866Z ##[section]Starting: Copy Files
2022-01-04T17:51:34.2762104Z ==============================================================================
2022-01-04T17:51:34.2762469Z Task         : Azure file copy
2022-01-04T17:51:34.2762789Z Description  : Copy files to Azure Blob Storage or virtual machines
2022-01-04T17:51:34.2763069Z Version      : 4.195.0
2022-01-04T17:51:34.2763324Z Author       : Microsoft Corporation
2022-01-04T17:51:34.2763679Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/azure-file-copy
2022-01-04T17:51:34.2764090Z ==============================================================================
2022-01-04T17:51:35.6812335Z ##[command]Import-Module -Name C:\Modules\az_6.5.0\Az.Accounts\2.7.0\Az.Accounts.psd1 -Global
2022-01-04T17:51:36.9280390Z ##[command]Clear-AzContext -Scope CurrentUser -Force -ErrorAction SilentlyContinue
2022-01-04T17:51:37.9678495Z ##[command]Clear-AzContext -Scope Process
2022-01-04T17:51:38.6841511Z ##[command]Connect-AzAccount -ServicePrincipal -Tenant *** -Credential System.Management.Automation.PSCredential -Environment AzureCloud @processScope
2022-01-04T17:51:40.3167982Z ##[command] Set-AzContext -SubscriptionId 61e25dca-a55a-45ee-9a73-80b5b16e99a2 -TenantId ***
2022-01-04T17:51:41.7602777Z ##[command]Import-Module -Name C:\Modules\az_6.5.0\Az.Resources\4.4.0\Az.Resources.psd1 -Global
2022-01-04T17:51:42.8948128Z ##[command]Import-Module -Name C:\Modules\az_6.5.0\Az.Storage\3.12.0\Az.Storage.psd1 -Global
2022-01-04T17:51:43.3003941Z ##[command]Import-Module -Name C:\Modules\az_6.5.0\Az.Compute\4.17.1\Az.Compute.psd1 -Global
2022-01-04T17:51:44.8576897Z ##[command]Import-Module -Name C:\Modules\az_6.5.0\Az.Network\4.12.0\Az.Network.psd1 -Global
2022-01-04T17:51:52.3485872Z ##[command] & "AzCopy\AzCopy.exe" login --service-principal --application-id "***" --tenant-id="***" --aad-endpoint "https://login.windows.net/"
2022-01-04T17:51:52.3877133Z INFO: If you set an environment variable by using the command line, that variable will be readable in your command line history. Consider clearing variables that contain credentials from your command line history.  To keep variables from appearing in your history, you can use a script to prompt the user for their credentials, and to set the environment variable.
2022-01-04T17:51:52.5378968Z INFO: AzCopy.exe: A newer version 10.13.0 is available to download
2022-01-04T17:51:52.5379412Z 
2022-01-04T17:51:52.6670910Z INFO: SPN Auth via secret succeeded.
2022-01-04T17:51:52.7774150Z INFO: AzCopy.exe: A newer version 10.13.0 is available to download
2022-01-04T17:51:52.7776936Z 
2022-01-04T17:51:52.7859690Z Uploading files from source path: 'D:\a\1\s' to storage account: 'armdeployfilestorage' in container: '30fbf4d0-2a75-4e2f-a8a7-a77aa4a9d9c7' with blob prefix: ''
2022-01-04T17:51:52.7957640Z ##[command] & "AzCopy\AzCopy.exe" copy "D:\a\1\s" "https://armdeployfilestorage.blob.core.windows.net/30fbf4d0-2a75-4e2f-a8a7-a77aa4a9d9c7"  --log-level=INFO --recursive
2022-01-04T17:51:52.8357960Z INFO: Scanning...
2022-01-04T17:51:52.8359470Z INFO: Authenticating to destination using Azure AD
2022-01-04T17:51:53.0512443Z INFO: AzCopy.exe: A newer version 10.13.0 is available to download
2022-01-04T17:51:53.0513192Z 
2022-01-04T17:51:53.0880420Z INFO: Any empty folders will not be processed, because source and/or destination doesn't have full folder support
2022-01-04T17:51:53.0948435Z 
2022-01-04T17:51:53.0974649Z Job 392de216-ebd4-8b4b-67a2-c3ce235da969 has started
2022-01-04T17:51:53.0976731Z Log file is located at: C:\Users\VssAdministrator\.azcopy\392de216-ebd4-8b4b-67a2-c3ce235da969.log
2022-01-04T17:51:53.0978594Z 
2022-01-04T17:51:53.1011154Z 
2022-01-04T17:51:55.0931441Z 0.0 %, 0 Done, 0 Failed, 83 Pending, 0 Skipped, 83 Total, 
2022-01-04T17:51:57.0933617Z 7.5 %, 81 Done, 0 Failed, 2 Pending, 0 Skipped, 83 Total, 2-sec Throughput (Mb/s): 793.5969
2022-01-04T17:51:59.0936567Z 51.1 %, 82 Done, 0 Failed, 1 Pending, 0 Skipped, 83 Total, 2-sec Throughput (Mb/s): 1444.9781
2022-01-04T17:52:01.0946692Z 97.8 %, 82 Done, 0 Failed, 1 Pending, 0 Skipped, 83 Total, 2-sec Throughput (Mb/s): 840.9941 
2022-01-04T17:52:01.0950890Z 
2022-01-04T17:52:01.0951510Z 
2022-01-04T17:52:01.0952232Z Job 392de216-ebd4-8b4b-67a2-c3ce235da969 summary
2022-01-04T17:52:01.0952754Z Elapsed Time (Minutes): 0.1334
2022-01-04T17:52:01.0953213Z Number of File Transfers: 83
2022-01-04T17:52:01.0953703Z Number of Folder Property Transfers: 0
2022-01-04T17:52:01.0954185Z Total Number of Transfers: 83
2022-01-04T17:52:01.0954661Z Number of Transfers Completed: 83
2022-01-04T17:52:01.0955134Z Number of Transfers Failed: 0
2022-01-04T17:52:01.0955574Z Number of Transfers Skipped: 0
2022-01-04T17:52:01.0956543Z TotalBytesTransferred: 770165203
2022-01-04T17:52:01.0958237Z Final Job Status: Completed
2022-01-04T17:52:01.0960772Z 
2022-01-04T17:52:01.2398620Z Uploaded files successfully from source path: 'D:\a\1\s' to storage account: 'armdeployfilestorage' in container: '30fbf4d0-2a75-4e2f-a8a7-a77aa4a9d9c7' with blob prefix: ''
2022-01-04T17:52:01.2411793Z ##[command] & "AzCopy\AzCopy.exe" logout
2022-01-04T17:52:01.2708972Z INFO: Logout succeeded.
2022-01-04T17:52:01.4340413Z INFO: AzCopy.exe: A newer version 10.13.0 is available to download
2022-01-04T17:52:01.4343249Z 
2022-01-04T17:52:01.4344082Z INFO: AzCopy.exe: A newer version 10.13.0 is available to download
2022-01-04T17:52:01.4345093Z 
2022-01-04T17:52:08.1438358Z [Azure Call]Getting the custom script extension 'WinRMCustomScriptExtension' for vm 'swordvmcertific'
2022-01-04T17:52:09.1974444Z [Azure Call]Setting the custom script extension 'WinRMCustomScriptExtension' for vm 'swordvmcertific'
2022-01-04T17:53:11.6823169Z [Azure Call]Set the custom script extension 'WinRMCustomScriptExtension' for vm 'swordvmcertific'
2022-01-04T17:53:11.9397238Z [Azure Call]Getting the status for vm 'swordvmcertific'
2022-01-04T17:53:12.8578801Z [Azure Call]Got the status for vm 'swordvmcertific'
2022-01-04T17:54:52.1830507Z ##[warning]A parameter cannot be found that matches parameter name 'and'.
2022-01-04T17:54:52.2781308Z ##[error]The remote session query failed for 13.87.91.35 with the following error message: The server certificate on the destination computer (13.87.91.35:5986) has the following errors:      
The SSL certificate is expired.    
2022-01-04T17:54:52.3372826Z ##[section]Finishing: Copy Files
## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs
2022-01-04T17:54:52.2781308Z ##[error]The remote session query failed for 13.87.91.35 with the following error message: The server certificate on the destination computer (13.87.91.35:5986) has the following errors:      
The SSL certificate is expired.
![ConfigureWinRM](https://user-images.githubusercontent.com/34490856/148143414-32217eee-e92a-470a-a684-757d42ca0ab9.PNG)

**Comment by marcusrichards** [comment]
This is our task script for info
`- task: AzureFileCopy@4
  displayName: Copy Files
  inputs:
    SourcePath: $(System.DefaultWorkingDirectory)
    azureSubscription: Sword GRC Paid Subscription (61e25dca-a55a-45ee-9a73-80b5b16e99a2)
    Destination: AzureVMs
    storage: armdeployfilestorage
    resourceGroup: ${{ parameters.ResourceGroup }}
    MachineNames: ${{ parameters.VmName }}
    vmsAdminUserName: ${{ parameters.VmUserName }}
    vmsAdminPassword: ${{ parameters.VmPassword }}
    TargetPath: d:\install
    enableCopyPrerequisites: true`

**Comment by aaroncoderepo** [comment]
We are also encountering the same issue with AzureFileCopy version 3.
This has a high impact as we are currently unable to use our DevOps pipelines to deploy.

Is anyone able to suggest any temporary work arounds?

**Comment by marcusrichards** [comment]
> We are also encountering the same issue with AzureFileCopy version 3. This has a high impact as we are currently unable to use our DevOps pipelines to deploy.
> 
> Is anyone able to suggest any temporary work arounds?

I found that pre-creating the self-cert certificate enables the AzureFileCopy to continue without making the new invalid certificate. You have to be careful to get the certificate spec right - this worked for me, must be run on a clean VM which has not already attempted to install WinRM scripting: 
`
$script = "New-SelfSignedCertificate -DnsName $(IpAddress), $(IpAddress) -TextExtension @('2.5.29.37={text}1.3.6.1.5.5.7.3.1') -KeyAlgorithm RSA -KeyLength 1024 -KeySpec Signature"
...
az vm run-command invoke  --command-id RunPowerShellScript ....vm auth...  --scripts $script
`

**Comment by aaroncoderepo** [comment]
Thanks for the suggestion Marcus.
Setting up clean VMs for our environments may be a problem as we have other software on the machines that is not scripted, so the process could be time consuming.

Hopefully we can find a solution for a VM with an existing certificate.

So far we have tried deleting the certificate and generating a new one using the makecrt command seen in the PowerShell script. After doing that we are getting the deployment error:

**Encountered an internal error in the SSL library.**

**Comment by marcusrichards** [comment]
I had the same error, internal error, which seems to be the WinRM script rerunning. In our case we can wipe it all and start again. There is a way to uninstall the WinRM script completely but I didn’t have to try that. Hope Microsoft pick this up quickly, hope you find a work around.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by stuartdotnet** [comment]
Also having the same issue with V3 and the AzureFileCopy step and unable to use any of the suggested workarounds. This is a critical disruption for our pipelines, can this be escalated?

**Comment by aaroncoderepo** [comment]
We managed to generate a new certificate running the cmd below.
Then configured WinRM to use the new certificate thumbprint by deleting the existing WinRM https configuration, and creating a new one.

Code was sourced from the following repo, and dates modified:

https://stackoverflow.com/questions/34897260/how-to-configure-winrm-for-azure-virtual-machine-in-resource-manager-portal

https://github.com/microsoft/azure-pipelines-extensions/blob/master/TaskModules/powershell/WinRM/WinRM-Http-Https-With-Makecert/ConfigureWinRM.ps1


Cmd Prompt

```
$hostname = "<FQDN>"

.\makecert -r -pe -n CN=$hostname -b 01/01/2022 -e 01/01/2032 -eku 1.3.6.1.5.5.7.3.1 -ss my -sr localmachine -sky exchange -sp "Microsoft RSA SChannel Cryptographic Provider" -sy 12

```
PowerShell

```
PS C:\Users\<user>> $hostname = "<FQDN>"

PS C:\Users\<user>> $thumbprint = (Get-ChildItem cert:\LocalMachine\My | Where-Object { $_.Subject -eq "CN=" + $hostname } | Select-Object -Last 1).Thumbprint

PS C:\Users\<user>> function Delete-WinRMListener
{
    $config = winrm enumerate winrm/config/listener
    foreach($conf in $config)
    {
        if($conf.Contains("HTTPS"))
        {
            Write-Verbose -Verbose "HTTPS is already configured. Deleting the exisiting configuration."

            winrm delete winrm/config/Listener?Address=*+Transport=HTTPS
            break
        }
    }
}

PS C:\Users\<user>> Delete-WinRMListener
VERBOSE: HTTPS is already configured. Deleting the exisiting configuration.

PS C:\Users\<user>> $WinrmCreate= "winrm create --% winrm/config/Listener?Address=*+Transport=HTTPS @{Hostname=`"$hostName`";CertificateThumbprint=`"$thumbPrint`"}"
    invoke-expression $WinrmCreate
    winrm set winrm/config/service/auth '@{Basic="true"}'

ResourceCreated
    Address = http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous
    ReferenceParameters
        ResourceURI = http://schemas.microsoft.com/wbem/wsman/1/config/listener
        SelectorSet
            Selector: Address = *, Transport = HTTPS

Auth
    Basic = true
    Kerberos = true
    Negotiate = true
    Certificate = false
    CredSSP = false
    CbtHardeningLevel = Relaxed

```

We have run this on our test environment and it seems to work.
Not sure if this is a good idea however, and there would be more elegant ways of accomplishing the same thing.
Possibly just modifying and running this PowerShell script:

https://github.com/microsoft/azure-pipelines-extensions/blob/master/TaskModules/powershell/WinRM/WinRM-Http-Https-With-Makecert/ConfigureWinRM.ps1

**Comment by meyemh30** [comment]
anyone else try updating the script for another 10 years? We need this to work.

**Comment by aaroncoderepo** [comment]
The fix will need to consider VMs with expiring/expired certificates:
- ensure that the certificate is deleted, and a new certificate is created (when the current certificate is about to expire or has expired)
- remove the old WinRM https config, recreate with the new certificate thumbprint

**Comment by mrsquish** [comment]
Seriously, how long does it take to review such a simple pull request.

**Comment by v-ibshaik** [comment]
We are able to Repro issue  in our local machine .Now we are updating code changes.

**Comment by thezaria** [comment]
Got the same issue, waiting for the fix

**Comment by v-ibshaik** [comment]
Hi All,  Could you please run the below work around for Temporary fix. Soon we will  raise the PR.

 The below script has to be run in  Target Machine.
 Download makecert.exe from " https://aka.ms/vstsmakecertexe" and Need to  save the file in PowerShell location.
 
Please Make Sure run three scripts in one ISE window:
 
 1. Create New Cert:

$hostname = hostname

$thumbprint = (Get-ChildItem cert:\LocalMachine\My | Where-Object { $_.Subject -eq "CN=" + $hostname } | Select-Object -Last 1).Thumbprint

if(-not $thumbprint)
{
if(-not (Test-Path -Path .\makecert.exe))
{
throw "File not found: makecert.exe"
}
Write-Host "Hello world1"
.\makecert -r -pe -n CN=$hostname -b (Get-Date).ToString('MM/dd/yyyy') -e (Get-Date).Addyears(10).ToString('MM/dd/yyyy') -eku 1.3.6.1.5.5.7.3.1 -ss my -sr localmachine -sky exchange -sp "Microsoft RSA SChannel Cryptographic Provider" -sy 12
$thumbprint=(Get-ChildItem cert:\Localmachine\my | Where-Object { $_.Subject -eq "CN=" + $hostname } | Select-Object -Last 1).Thumbprint

if(-not $thumbprint)
{
throw "Failed to create the test certificate."
}
}

2. Delete WinRM Listner:

$hostname = hostname

$thumbprint = (Get-ChildItem cert:\LocalMachine\My | Where-Object { $_.Subject -eq "CN=" + $hostname } | Select-Object -Last 1).Thumbprint

function Delete-WinRMListener
{
    $config = winrm enumerate winrm/config/listener
    foreach($conf in $config)
    {
        if($conf.Contains("HTTPS"))
        {
            Write-Verbose -Verbose "HTTPS is already configured. Deleting the exisiting configuration."
winrm delete winrm/config/Listener?Address=*+Transport=HTTPS
            break
        }
    }
}

Delete-WinRMListener


3. WinRM set with new cert 

$WinrmCreate= "winrm create --% winrm/config/Listener?Address=*+Transport=HTTPS @{Hostname=`"$hostName`";CertificateThumbprint=`"$thumbPrint`"}"
invoke-expression $WinrmCreate
winrm set winrm/config/service/auth '@{Basic="true"}'

Thanks.

**Comment by meyemh30** [comment]
I was able to test this and confirmed it worked @v-ibshaik . Here is the updated powershell script and instructions, make sure to use this on the target VM. This is a workaround, I've also created a Severity A support ticket with Microsoft and their Product Team is pushing a patch into place. Until then, use this:


# 1: download makecert put in same directory as powershell script: https://aka.ms/vstsmakecertexe
# 2: copy paste script into ADMIN powershell ISE
# 3: Change the below variable
# 4: Open up certificates.msc and delete your personal store, expired certificate on the TARGET VM
# 5: Run script

$hostname = '<USE IP or FQDN hostname>'

$thumbprint = (Get-ChildItem cert:\LocalMachine\My | Where-Object { $_.Subject -eq "CN=" + $hostname } | Select-Object -Last 1).Thumbprint

if(-not $thumbprint)
{
    if(-not (Test-Path -Path .\makecert.exe))
    {
        throw "File not found: makecert.exe"
    }
    .\makecert -r -pe -n CN=$hostname -b (Get-Date).ToString('MM/dd/yyyy') -e (Get-Date).Addyears(10).ToString('MM/dd/yyyy') -eku 1.3.6.1.5.5.7.3.1 -ss my -sr localmachine -sky exchange -sp "Microsoft RSA SChannel Cryptographic Provider" -sy 12
    $thumbprint=(Get-ChildItem cert:\Localmachine\my | Where-Object { $_.Subject -eq "CN=" + $hostname } | Select-Object -Last 1).Thumbprint

    if(-not $thumbprint)
    {
        throw "Failed to create the test certificate."
    }
} 

# Delete all existing win rm listener
function Delete-WinRMListener
{
    $config = winrm enumerate winrm/config/listener
    foreach($conf in $config)
    {
        if($conf.Contains("HTTPS"))
        {
            Write-Verbose -Verbose "HTTPS is already configured. Deleting the exisiting configuration."
            winrm delete winrm/config/Listener?Address=*+Transport=HTTPS
            break
        }
    }
}

Delete-WinRMListener 



# Recreate WinRM listener with new certificate created above
$WinrmCreate= "winrm create --% winrm/config/Listener?Address=*+Transport=HTTPS @{Hostname=`"$hostName`";CertificateThumbprint=`"$thumbPrint`"}"
    invoke-expression $WinrmCreate
    winrm set winrm/config/service/auth '@{Basic="true"}'

**Comment by v-ibshaik** [comment]
Hi All,

We have fixed the issue.

Thanks.

**Comment by marcusrichards** [comment]
great news thanks. Has it been deployed to the production environment?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by v-ibshaik** [comment]
Hello  marcusrichards,

Yes, We have fixed the issue by updating the Cert Valid dates.

Thanks.

**Comment by marcusrichards** [comment]
Good news, thank you.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jelsner-nlp** [comment]
Hello v-ibshaik,
I am confused here.  Does this mean if we are using the AzureFileCopy task in an AzureDevops pipeline, should error be fixed or do we have to do anything on the target servers to update the certs still?

**Comment by mrsquish** [comment]
I'm not seeing any fix on my release pipelines.

**Comment by jelsner-nlp** [comment]
Me neither..

**Comment by v-ibshaik** [comment]
Hi mrsquish, 

We don't have any code changes in azure-pipeline-tasks and basically we have updated the code changes  in ==> azure-quickstart-templates/master/demos/vm-winrm-windows/ConfigureWinRM.ps1
Could you please verify the task .

Thanks.

**Comment by mrsquish** [comment]
Thanks v-ibshaik.

So, I assumed I was supposed to run the ConfigureWinRM.ps1 script on the VM, which I did. Using the DNS name as the hostname? After that I ran the release pipeline, this time getting the dreaded "Operation 'PUT Extension' is not allowed on VM extension 'WinRMCustomScriptExtension' since it is marked for deletion.". Tried, restarting the VM (no change). I suspected the restart didn't work, so tried restarting from windows. Now the VM is completely unresponsive, can't start it, can't stop it, can't connect to it. Portal thinks it's running, but my guess is it isn't.

**Comment by mrsquish** [comment]
Ok, so I was able to resurrect the VM with a restart --force. 
After doing so, the WinRMCustomScript extension had been removed, so I thought, good to go. 
Ran the Release pipeline again, and in short, it wouldn't finish (end up stopping it after the AzureCopy task had been running for an hour). 
The last two lines from the log was :
[debug]Set-AzVMCustomScriptExtension -ResourceGroupName <REDACTED> -VMName <REDACTED> -Name WinRMCustomScriptExtension -FileUri https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/demos/vm-winrm-windows/ConfigureWinRM.ps1 https://raw.githubusercontent.com/Azure/azure-quickstart-templates/501dc7d24537e820df7c80bce51aba9674233b2b/201-vm-winrm-windows/makecert.exe  -Run ConfigureWinRM.ps1 -Argument <REDACTED>.australiaeast.cloudapp.azure.com -Location australiaeast -ErrorAction Stop -Verbose
[debug]Performing the operation "Set" on target "CustomScriptExtension".

Could it be possible the argument isn't being passed properly and it's waiting for user input?

**Comment by v-ibshaik** [comment]
Hello mrsquish,

1. Issue got fixed ,No need to do any update  in Target machine. could you please Try to connect new Target machine from your Task.

2. In your case you have work around with  your target machine. for this you need to  delete the cert ,run the Delete-WinRMListener and clear the cache.

Thanks.

**Comment by themattconnolly** [comment]
In case anyone stumbles on this thread, I had a Windows 2012 R2 VM in Azure created maybe in 2016, with an Azure file copy task in Azure DevOps that started failing due to certificates that had expired in 2021. Updating the task version from 1.* to 4.* and running the release again had no effect on the certificate. I ended up following the script from @[meyemh30](https://github.com/meyemh30)to delete and recreate the certificate. I also had to manually download AzCopy and add the path to the system environment variables. I also had to adjust my source path to add "\**" to the end to make sure the source folder name wasn't created and used as the destination. My specific challenges may have been unique to my combination, but I'm back up and running again.
